Index: tools/clang/lib/Parse/ParseDecl.cpp
===================================================================
--- tools/clang/lib/Parse/ParseDecl.cpp	(revision unknown)
+++ tools/clang/lib/Parse/ParseDecl.cpp	(working copy)
@@ -2978,14 +2978,22 @@
       //   static const bool __is_signed;
       //
       // then treat __is_signed as an identifier rather than as a keyword.
-      if (DS.getTypeSpecType() == TST_bool &&
-          DS.getTypeQualifiers() == DeclSpec::TQ_const &&
-          DS.getStorageClassSpec() == DeclSpec::SCS_static)
-        TryKeywordIdentFallback(true);
-
-      // We're done with the declaration-specifiers.
-      goto DoneWithDeclSpec;
-
+      switch (DS.getTypeSpecType()) {
+      case TST_typename:
+        {
+          auto SeenType = Sema::GetTypeFromParser(DS.getRepAsType());
+          if (!SeenType->isBooleanType())
+            goto DoneWithDeclSpec;
+          // fall through
+        }
+      case TST_bool:
+        if (DS.getTypeQualifiers() == DeclSpec::TQ_const &&
+            DS.getStorageClassSpec() == DeclSpec::SCS_static)
+          TryKeywordIdentFallback(true);
+        // fall through
+      default:
+        goto DoneWithDeclSpec;
+      }
       // typedef-name
     case tok::kw___super:
     case tok::kw_decltype:
